### ISSUE TITLE
About the training time

### ISSUE THREAD

**Issue by WentangChen** [EasyRy/CRA-PCN#11]
你好，非常感谢你能公开论文代码！
我想问一下训练pcn，shapenet55，shapenet34，mvp这四个数据集大概在多少epoch的时候会得到论文的结果呀？

**Comment by EasyRy** [comment]
您好，感谢关注。
1. PCN和MVP可以参考下提供的log。PCN我们训练400个epoch，看log的话大概300左右就接近收敛了，两张V100开满的话大概3天。MVP训50个epoch就能收敛。
2. ShapeNet55和34我差不多训练了300个epoch，大概需要一周多时间。

**Comment by WentangChen** [comment]
V100是16G的显存吗

**Comment by EasyRy** [comment]
32G

**Comment by WentangChen** [comment]
请问怎么开多卡训练呢？我设置了
os.environ['CUDA_VISIBLE_DEVICES'] = '4,5,6,7'
和
__C.CONST.DEVICE                                 = '4,5,6,7'
运行脚本python train_pcn.py 一直没反应

**Comment by EasyRy** [comment]
请问没反应指的是不能在指定的卡上训练？还是不能启动训练？

**Comment by WentangChen** [comment]
不能启动训练，一直卡住：
![image](https://github.com/user-attachments/assets/4370c388-eba9-4f34-94e7-bfd0efcf558f)
![image](https://github.com/user-attachments/assets/10dc5520-1c30-48d2-b8a2-2e122ce85776)

**Comment by EasyRy** [comment]
我没碰到过这个问题哎...你可以试试把lr_scheduler.step的位置([here](https://github.com/EasyRy/CRA-PCN/blob/main/manager.py#L99))挪到每个epoch的结尾。或者打印一些log，看看卡在哪？

**Comment by WentangChen** [comment]
lr_scheduler.step我试了没有用，请问一般用什么打印log呢？

**Comment by EasyRy** [comment]
就简单print吧。

这里的训练代码和seedformer与snowflakenet大差不差的，你可以试试看他们的训练能不能跑起来。

你用训练好的权重推理过吗，能跑成功吗？

**Comment by WentangChen** [comment]
我用训练好的权重推理，已经跑成功过。
seedformer和snowflakenet的多GPU也跑不了，我在想是不是环境不一样的原因

**Comment by EasyRy** [comment]
这我就不太清楚了…或许可以用其他你能跑通的训练代码来跑这个model

**Comment by WentangChen** [comment]
你好，我想问一下：训练shapnet时，Testing scores指的是？

**Comment by WentangChen** [comment]
你好，我想问一下，我在shapenet55训练好之后的模型的CD-S是0.50，CH-M是0.77，CD-H是1.58。后面两项与论文差别比较大，请问是什么原因呢？

**Comment by EasyRy** [comment]
你好，
1. testing score指的就是测试时的CD值。
2. 确实差的比较多，能看一下训练的log吗？然后，用的seed generator是哪一个呢？

**Comment by WentangChen** [comment]
1. 是CD-L1-avg吗？
2. 是提供training.txt吗？seed generator用的是SeedGenerator

**Comment by EasyRy** [comment]
1. 是的。
2. testing.txt, 这是我训练得到的文件。你可以参考下。
[testing.txt](https://github.com/user-attachments/files/17413333/testing.txt)


另外附上我当时的测试结果，[easy.txt](https://github.com/user-attachments/files/17413354/easy.txt)，[median.txt](https://github.com/user-attachments/files/17413356/median.txt)，[hard.txt](https://github.com/user-attachments/files/17413355/hard.txt)

ps. 权重的话我也有，但是代码重构过，所以load不了...

**Comment by WentangChen** [comment]
1.这是训练的log
[training.txt](https://github.com/user-attachments/files/17414869/training.txt)
因为我用的事分布式训练，所有有一点点乱。
2.请问还有重构之前的代码吗？我训练了很多次shapenet55，testing scores收敛在11.4左右，但是CD-S，CH-M，CD-H还是和上述一样差别很大；而在训练pcn的时候cd能在6.44这样子。请问差别比较大的原因是什么呢？

**Comment by WentangChen** [comment]
抱歉叨扰你了
1. 我训练PCN的时候，显示#400 27.7835 19.3694 11.6523 27.7835 | 6.4118 | #331 6.3915，但是我测试test.py时都是6.44 6.43左右，请问是什么原因呢？

**Comment by EasyRy** [comment]
1. 有的，你留个邮箱。
2. train.py训练代码里的测试代码和test.py一样吧。感觉有点波动也正常？

**Comment by WentangChen** [comment]
1. 1099049741@qq.com，非常感谢！
2. 请问有没有办法设置训练时和测试时的结果保持一致？
3. 我还是想问问关于 testing scores收敛在11.4左右，但是CD-S，CH-M，CD-H还是差别比较大的原因会有哪些呀？

**Comment by EasyRy** [comment]
你好，邮件已发送。关于问题2，固定下随即种子看看？
关于问题3，也许是推理的时候有数值溢出了，可以调大插值时候的常数值。

**Comment by WentangChen** [comment]
1. 十分感谢~~
2. 我在train_pcn.py的主函数里设置了：
`    seed = 2024
    set_seed(seed)`
   这样是算作固定吗？
3. 请问在哪里调大插值的常数值呢？
十分感谢你的悉心回答~

**Comment by EasyRy** [comment]
2. 算吧。这样每次跑出来结果应该都一样吧。
3. https://github.com/EasyRy/CRA-PCN/tree/main/models#tips

**Comment by the-king-of-yhq** [comment]
> 请问怎么开多卡训练呢？我设置了 os.environ['CUDA_VISIBLE_DEVICES'] = '4,5,6,7' 和 __C.CONST.DEVICE = '4,5,6,7' 运行脚本python train_pcn.py 一直没反应

你好，请问你这问题解决了吗，我无论怎么设置这个os.environ['CUDA_VISIBLE_DEVICES']都没用，都会默认分布，而且还是不均匀的

**Comment by the-king-of-yhq** [comment]
![image](https://github.com/user-attachments/assets/566f6216-2743-46ea-8665-9d3e2cf1e7d5)

**Comment by WentangChen** [comment]
> > 请问怎么开多卡训练呢？我设置了 os.environ['CUDA_VISIBLE_DEVICES'] = '4,5,6,7' 和 __C.CONST.DEVICE = '4,5,6,7' 运行脚本python train_pcn.py 一直没反应
> 
> 你好，请问你这问题解决了吗，我无论怎么设置这个os.environ['CUDA_VISIBLE_DEVICES']都没用，都会默认分布，而且还是不均匀的

我修改代码设置了分布式训练，跑起来的时候是可以均匀分布的

**Comment by the-king-of-yhq** [comment]
> > > 请问怎么开多卡训练呢？我设置了 os.environ['CUDA_VISIBLE_DEVICES'] = '4,5,6,7' 和 __C.CONST.DEVICE = '4,5,6,7' 运行脚本python train_pcn.py 一直没反应
> > 
> > 
> > 你好，请问你这问题解决了吗，我无论怎么设置这个os.environ['CUDA_VISIBLE_DEVICES']都没用，都会默认分布，而且还是不均匀的
> 
> 我修改代码设置了分布式训练，跑起来的时候是可以均匀分布的

你好如何修改的，可以发下吗？

**Comment by WentangChen** [comment]
> > > > 请问怎么开多卡训练呢？我设置了 os.environ['CUDA_VISIBLE_DEVICES'] = '4,5,6,7' 和 __C.CONST.DEVICE = '4,5,6,7' 运行脚本python train_pcn.py 一直没反应
> > > 
> > > 
> > > 你好，请问你这问题解决了吗，我无论怎么设置这个os.environ['CUDA_VISIBLE_DEVICES']都没用，都会默认分布，而且还是不均匀的
> > 
> > 
> > 我修改代码设置了分布式训练，跑起来的时候是可以均匀分布的
> 
> 你好如何修改的，可以发下吗？

你留一下邮箱，我去实验室的时候copy发给你

**Comment by EasyRy** [comment]
Close this issue since no more response. Feel free to reopen it if you have more questions.

**Comment by zzyykk725** [comment]
> > > > > 请问怎么开多卡训练呢？我设置了 os.environ['CUDA_VISIBLE_DEVICES'] = '4,5,6,7' 和 __C.CONST.DEVICE = '4,5,6,7' 运行脚本python train_pcn.py 一直没反应
> > > > 
> > > > 
> > > > 你好，请问你这问题解决了吗，我无论怎么设置这个os.environ['CUDA_VISIBLE_DEVICES']都没用，都会默认分布，而且还是不均匀的
> > > 
> > > 
> > > 我修改代码设置了分布式训练，跑起来的时候是可以均匀分布的
> > 
> > 
> > 你好如何修改的，可以发下吗？
> 
> 你留一下邮箱，我去实验室的时候copy发给你

你好，请问怎么设置多卡训练，我设置的DDP跑起来LOSS值非常大, 开始时候几千上万了，感觉很不正常，单卡跑都几十，麻烦请发下多卡训练的代码设置，1808472832@qq.com
